### PR TITLE
Add support of new type consumers to cloning fanout connectors

### PR DESCRIPTION
**Pull request by dmitryax**
**Description:** Cloning fanout connectors now able to work with new type of downstream consumers

**Testing:** Unit tests

Aslo added MetricData generators to `internal/data/testdata/metric.go` 
Code in `internal/data/testdata/common.go` moved from `internal/data/testdata/trace.go` as is

### CODE REVIEW

**Comment by bogdandrutu** [comment]
Please use https://github.com/open-telemetry/opentelemetry-collector/pull/718 as inspiration for the metrics testdata. I will drop that PR once you copy the metrics testdata.

**Comment by dmitryax** [comment]
> Please use #718 as inspiration for the metrics testdata. I will drop that PR once you copy the metrics testdata.

done

File: processor/cloningfanoutconnector.go [hunk 1]
```diff
@@ -35,37 +38,44 @@ import (
...
-	for _, baseMetricsConsumer := range baseMetricsConsumers {
-		metricsConsumer, ok := baseMetricsConsumer.(consumer.MetricsConsumerOld)
-		if !ok {
-			panic("CreateMetricsCloningFanOutConnector does not support new type of MetricsConsumer")
+func CreateMetricsCloningFanOutConnector(mcs []consumer.MetricsConsumerBase) consumer.MetricsConsumerBase {
+	metricsConsumersOld := make([]consumer.MetricsConsumerOld, 0, len(mcs))
+	metricsConsumers := make([]consumer.MetricsConsumer, 0, len(mcs))
+	allMetricsConsumersOld := true
+	for _, mc := range mcs {
+		if metricsConsumer, ok := mc.(consumer.MetricsConsumer); ok {
```
Comment (pjanotti) [inline processor/cloningfanoutconnector.go]: nit: because if/else and another type switch with a local defined on `switch` probably will look simpler

File: internal/data/testdata/trace.go [hunk 2]
```diff
@@ -228,15 +228,15 @@ func generateTraceDataTwoSpansSameResourceOneDifferent() data.TraceData {
 	fillSpanOne(rs0ils0.Spans().Get(0))
 	fillSpanTwo(rs0ils0.Spans().Get(1))
 	rs1 := td.ResourceSpans().Get(1)
-	fillResource2(rs1.Resource())
+	rs1.Resource().SetAttributes(data.NewAttributeMap(resourceAttributes2))
```
Comment (bogdandrutu) [inline internal/data/testdata/trace.go]: Why change?

File: internal/data/testdata/trace.go [hunk 3]
```diff
@@ -94,27 +94,27 @@ var (
...
 	}
 }
 
-func generateTraceDataNoLibraries() data.TraceData {
-	td := generateTraceDataOneEmptyResourceSpans()
+func GenerateTraceDataNoLibraries() data.TraceData {
+	td := GenerateTraceDataOneEmptyResourceSpans()
 	rs0 := td.ResourceSpans().Get(0)
-	fillResource1(rs0.Resource())
+	rs0.Resource().SetAttributes(data.NewAttributeMap(resourceAttributes1))
```
Comment (bogdandrutu) [inline internal/data/testdata/trace.go]: Why this change?

File: processor/cloningfanoutconnector.go (hunk 1 above)
Comment (bogdandrutu) [reply processor/cloningfanoutconnector.go]: I personally like more if/else

File: processor/cloningfanoutconnector.go [hunk 4]
```diff
@@ -150,7 +233,16 @@ func cloneTraceData(td *consumerdata.TraceData) *consumerdata.TraceData {
 	return clone
 }
 
-func cloneMetricsData(md *consumerdata.MetricsData) *consumerdata.MetricsData {
+func cloneTraceData(td data.TraceData) data.TraceData {
+	resourceSpansClones := make([]*otlptrace.ResourceSpans, 0, len(data.TraceDataToOtlp(td)))
+	for _, resourceSpans := range data.TraceDataToOtlp(td) {
```
Comment (bogdandrutu) [inline processor/cloningfanoutconnector.go]: otlp := data.TraceDataToOtlp(td)

Just in case data.TraceDataToOtlp will become a bit more expensive.

File: processor/cloningfanoutconnector.go [hunk 5]
```diff
@@ -150,7 +233,16 @@ func cloneTraceData(td *consumerdata.TraceData) *consumerdata.TraceData {
 	return clone
 }
 
-func cloneMetricsData(md *consumerdata.MetricsData) *consumerdata.MetricsData {
+func cloneTraceData(td data.TraceData) data.TraceData {
```
Comment (bogdandrutu) [inline processor/cloningfanoutconnector.go]: I would add this "Clone" to the MetricData, just in case we will have more than the ResourceSpans list. Can be done in a separate PR.

File: processor/cloningfanoutconnector.go [hunk 6]
```diff
@@ -167,3 +259,12 @@ func cloneMetricsData(md *consumerdata.MetricsData) *consumerdata.MetricsData {
 
 	return clone
 }
+
+func cloneMetricsData(md data.MetricData) data.MetricData {
```
Comment (bogdandrutu) [inline processor/cloningfanoutconnector.go]: Same as cloneTraceData.

File: processor/cloningfanoutconnector.go (hunk 4 above)
Comment (dmitryax) [reply processor/cloningfanoutconnector.go]: done

File: internal/data/testdata/trace.go (hunk 3 above)
Comment (dmitryax) [reply internal/data/testdata/trace.go]: Messed up with rebase. restored

File: processor/cloningfanoutconnector.go (hunk 5 above)
Comment (dmitryax) [reply processor/cloningfanoutconnector.go]: moved

File: processor/cloningfanoutconnector.go (hunk 1 above)
Comment (dmitryax) [reply processor/cloningfanoutconnector.go]: This condition is only for the transition state. So will be removed soon anyway :)

**Comment by dmitryax** [comment]
@bogdandrutu @pjanotti PTAL again

File: internal/data/metric.go [hunk 7]
```diff
@@ -49,6 +50,17 @@ func NewMetricData() MetricData {
...
+func (md MetricData) Clone() MetricData {
+	otlp := MetricDataToOtlp(md)
+	resourceMetricsClones := make([]*otlpmetrics.ResourceMetrics, 0, len(otlp))
+	for _, resourceMetrics := range otlp {
+		resourceMetricsClones = append(resourceMetricsClones,
+			proto.Clone(resourceMetrics).(*otlpmetrics.ResourceMetrics))
+	}
+	return MetricDataFromOtlp(resourceMetricsClones)
+}
+
```
Comment (bogdandrutu) [inline internal/data/metric.go]: Some tests for clone would be good. Pass an otlp check that Clone returns equalValues but not same pointer.

Comment (dmitryax) [reply internal/data/metric.go]: it's being tested, but not this function directly https://github.com/open-telemetry/opentelemetry-collector/pull/717/files#diff-50214c26172368731479968d001a4102R222

File: processor/cloningfanoutconnector.go (hunk 1 above)
Comment (dmitryax) [reply processor/cloningfanoutconnector.go]: actually `switch` will not give us right result. we need only one behavior even if `ms` implements both interfaces, and that's what `if` provides
